### PR TITLE
HSEARCH-2781 Rework internals of sharding to make it possible for backends not to rely on Class<?>

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net//browse/HSEARCH-2781

After this patch, integrators will be able to write `entityIndexBinding.getIndexManagerSelector().forExisting(typeId, null, null)` to get all index managers containing a document for the given type. No use of `Class<?>` required.

### CODE REVIEW

File: engine/src/main/java/org/hibernate/search/indexes/impl/IndexShardingStrategyIndexManagerSelector.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	private final IndexManager[] allIndexManagers;
+
+	public IndexShardingStrategyIndexManagerSelector(IndexShardingStrategy shardingStrategy, IndexManager[] allIndexManagers) {
+		super();
+		this.shardingStrategy = shardingStrategy;
+		this.allIndexManagers = allIndexManagers;
+	}
+
+	@Override
+	public Set<IndexManager> all() {
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/indexes/impl/IndexShardingStrategyIndexManagerSelector.java]: The `Set` returned by this method could be initialized during constructor?

**Comment by Sanne** [comment]
merged. Thanks!

File: engine/src/main/java/org/hibernate/search/indexes/impl/IndexShardingStrategyIndexManagerSelector.java (hunk 1 above)
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/indexes/impl/IndexShardingStrategyIndexManagerSelector.java]: I see you did it; thanks.
It should work, yes, but I wasn't sure whether strategies were allowed to add/remove index managers at runtime or not. I guess they're not.

Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/indexes/impl/IndexShardingStrategyIndexManagerSelector.java]: Sure. That's the main difference between the dynamic sharding strategies and the "legacy" (this one)

**Pull request closed without merging**